mlflow/openai/api_request_parallel_processor.py

```diff
@@ -0,0 +1,349 @@
+            await asyncio.sleep(seconds_to_sleep_each_loop)
+
+            # if a rate limit error was hit recently, pause to cool down
+            seconds_since_rate_limit_error = (
+                time.time() - status_tracker.time_of_last_rate_limit_error
+            )
+            if seconds_since_rate_limit_error < seconds_to_pause_after_rate_limit_error:
+                remaining_seconds_to_pause = (
+                    seconds_to_pause_after_rate_limit_error - seconds_since_rate_limit_error
+                )
+                await asyncio.sleep(remaining_seconds_to_pause)
+                # ^e.g., if pause is 15 seconds and final limit was hit 5 seconds ago
+                cool_down_time = time.ctime(
+                    status_tracker.time_of_last_rate_limit_error
+                    + seconds_to_pause_after_rate_limit_error
+                )
+                _logger.warning(f"Pausing to cool down until {cool_down_time}")
+
+        # after finishing, log final status
+        if status_tracker.num_tasks_failed > 0:
+            raise mlflow.MlflowException(
+                f"{status_tracker.num_tasks_failed} tasks failed. See logs for details."
+            )
+        if status_tracker.num_rate_limit_errors > 0:
+            _logger.warning(
+                f"{status_tracker.num_rate_limit_errors} rate limit errors received. "
+                "Consider running at a lower rate."
+            )
+
+
+def run_as_subprocess(requests):
+    # Why not directly call `process_api_requests`?
+    # `process_api_requests` is a coroutine. To call it in a normal function,
+    #  we need to wrap it in `asyncio.run`, but `asyncio.run` cannot be called
+    # from a running event loop. This is a problem when running in Jupyter Notebook
+    # (see https://stackoverflow.com/q/55409641 for more details).
+    # To work around this, we call `process_api_requests` in a subprocess.
+    _logger.info(f"Processing {len(requests)} requests in parallel...")
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_dir = pathlib.Path(temp_dir)
+        input_file = temp_dir / "input.jsonl"
+        output_file = temp_dir / "output.jsonl"
+        with open(input_file, "w") as f:
+            for request in requests:
+                json.dump(request, f)
+                f.write("\n")
+        subprocess.run(
+            [
+                sys.executable,
+                __file__,
+                "--input-file",
+                input_file,
+                "--output-file",
+                output_file,
+            ],
+            check=True,
+        )
+        with open(output_file) as f:
+            responses = []
+            for line in f:
+                index, response = line.split(maxsplit=1)
+                index = int(index)
+                response = json.loads(response)
+                responses.append((index, response))
+            return [r[1] for r in sorted(responses, key=lambda x: x[0])]
```

Comment:
@jinzhang21 @BenWilson2 What do you think about this approach? Yes, input/output serialization/deserialization can be expensive, but this allows us to use async requests.

Reply:
I think doing something similar with threading is much more complicated than doing it with async.

Reply:
If we use threadpool approach, I think lots of existing code need to change, unfortunately. Essentially anything marked with `async`, `await` etc needs to be removed. The idea will be for each `APIRequest` class, create a thread, fire it with that thread, and join at the end for all requests in the main thread.
Something like this: https://stackoverflow.com/questions/57284126/how-can-i-use-threading-with-requests

Reply:
This is really a bummer tbh. Changing from async to sync approach needs some major changes. I consulted with GPT-4, here is its answer. Maybe you want to try and see if it works.

```
import argparse
import json
import pathlib
import sys
import subprocess
import tempfile
from dataclasses import dataclass
from concurrent.futures import ThreadPoolExecutor

import tiktoken
import openai
import openai.error

import mlflow
from mlflow.environment_variables import _MLFLOW_OPENAI_TESTING

_logger = logging.getLogger(__name__)


@dataclass
class StatusTracker:
    # ... (Same as before) ...


@dataclass
class APIRequest:
    # ... (Same as before, but without 'async' keyword and without 'await' keyword) ...

    def call_api(self, retry_queue: list, status_tracker: StatusTracker):
        """
        Calls the OpenAI API and stores results.
        """
        _logger.debug(f"Request #{self.index} started")
        try:
            response = openai.ChatCompletion.create(**self.request_json)
            _logger.debug(f"Request #{self.index} succeeded")
            status_tracker.complete_task(success=True)
            with open(self.output_file, "a") as f:
                f.write(f"{self.index} ")  # this will be used later to sort the responses
                json.dump(response, f)
                f.write("\n")
        except openai.error.RateLimitError as e:
            _logger.warning(f"Request #{self.index} failed with {e!r}")
            status_tracker.time_of_last_rate_limit_error = time.time()
            status_tracker.num_rate_limit_errors += 1
        # ... (Same as before, with only the necessary changes) ...


# ... (Same as before, for num_tokens_consumed_from_request function) ...


def process_api_requests(
    input_file: str,
    output_file: str,
    max_requests_per_minute: float = 1_500,
    max_tokens_per_minute: float = 125_000,
    token_encoding_name: str = "cl100k_base",
    max_attempts: int = 5,
    num_threads: int = 10,
):
    # ... (Same as before, with changes from async to sync approach) ...

    # `requests` will provide requests one at a time
    with open(input_file) as f:
        requests_iter = enumerate(f)
        with ThreadPoolExecutor(max_workers=num_threads) as executor:
            while True:
                # ... (Same as before, with changes from async to sync approach) ...

                # if enough capacity available, call API
                if next_request:
                    next_request_tokens = next_request.token_consumption
                    if (
                        available_request_capacity >= 1
                        and available_token_capacity >= next_request_tokens
                    ):
                        # update counters
                        available_request_capacity -= 1
                        available_token_capacity -= next_request_tokens
                        next_request.attempts_left -= 1

                        # call API
                        executor.submit(
                            next_request.call_api,
                            retry_queue=queue_of_requests_to_retry,
                            status_tracker=status_tracker,
                        )
                        next_request = None  # reset next_request to empty

                # ... (Same as before, with changes from async to sync approach) ...


# ... (Same as before, for run_as_subprocess function) ...
```

Reply:
thanks, let me check if this works and is thread-safe.

Reply:
` -= x` and `+= y` are not thread-safe. We need a lock.

Reply:
Updated